### DISCUSSION TITLE
How to report at most one diagnostic per line?

### DISCUSSION THREAD

**Discussion by MartyIX** [dotnet/roslyn#56316] [Q&A] [answered]
Hi,

I followed the tutorial here https://docs.microsoft.com/en-us/dotnet/csharp/roslyn-sdk/tutorials/how-to-write-csharp-analyzer-code-fix and I have an C# analyzer which works great. However, I wonder whether I can report only at most one diagnostic per a source file line. The first one ideally.

Is that possible?

Thank you!

**Comment by CyrusNajmabadi** [comment, marked as answer]
Yes.  Just have your analyzer keep track of the diagnostics it is producing, and skip future ones of they would be on the same line as one you already emitted.

**Reply by MartyIX** [reply]
But I don't really know when to start and stop tracking, do I? I mean lifecycle of diagnostic reports is unclear to me

**Reply by MartyIX** [reply]
The alternative of remembering everything does not work either as I don't know when a diagnostic is removed.

**Reply by MartyIX** [reply]
Hopefuly, it's clear what I mean. Obviously, I miss some useful callback.

**Reply by CyrusNajmabadi** [reply]
> But I don't really know when to start and stop tracking, do I?

```c#
public override void Initialize(AnalysisContext context)
{
    context.RegisterSyntaxTreeAction(ProcessTree);
}

private void ProcessTree(SyntaxTreeAnalysisContext context)
{
    var linesThatIveReportedADiagOn = new HashSet<int>();
    // do your work, but don't bother calling context.ReportDiagnostic for a diagnostic on a line you've already reported.
}
```

**Reply by CyrusNajmabadi** [reply]
> The alternative of remembering everything does not work either as I don't know when a diagnostic is removed.

We always just compute the full new set of diagnostics for your anlayzer, and hten replce the entire previous set with the freshly computed set.

**Reply by MartyIX** [reply]
Thanks!!
